torchtune/models/gemma/transformer.py

```diff
@@ -0,0 +1,111 @@
+        max_seq_len (int): maximum sequence length the model will be run with, as used
+            by :func:`~torchtune.modules.KVCache`
+        num_heads (int): number of query heads. For MHA this is also the
+            number of heads for key and value. This is used to setup the
+            :func:`~torchtune.modules.KVCache`
+        head_dim (int): embedding dimension for each head in self-attention. This is used
+            to setup the :func:`~torchtune.modules.KVCache`
+        norm (nn.Module): Callable that applies normalization to the output of the decoder,
+            before final MLP.
+        output (nn.Linear): Callable that applies a linear transformation to the output of
+            the decoder.
+        norm_embeddings (bool): Whether to apply normalization before the self-attention layer, defaults to False.
+
+    Note:
+        Arg values are checked for correctness (eg: ``attn_dropout`` belongs to [0,1])
+        in the module where they are used. This helps reduces the number of raise
+        statements in code and improves readability.
+    """
+
+    def __init__(
+        self,
+        tok_embeddings: nn.Embedding,
+        layer: TransformerDecoderLayer,
+        num_layers: int,
+        max_seq_len: int,
+        num_heads: int,
+        head_dim: int,
+        norm: nn.Module,
+        output: nn.Linear,
+        norm_embeddings: bool = False,
+    ) -> None:
+        super().__init__()
+        self.tok_embeddings = tok_embeddings
+        self.layers = _get_clones(layer, num_layers)
+        self.norm = norm
+        self.output = output
+        self.max_seq_len = max_seq_len
+        self.num_heads = num_heads
+        self.head_dim = head_dim
+        self.causal_mask = None
+        self.norm_embeddings = norm_embeddings
+
+    def forward(
+        self, tokens: Tensor, mask: Optional[Tensor] = None, curr_pos: int = None
+    ) -> Tensor:
+        """
+        Args:
+            tokens (Tensor): input tensor with shape [b x s]
+            mask (Optional[Tensor]): attention mask tensor, defaults to None.
+            curr_pos (int): current position in the seq, defaults to 0.
+                Only relevant when incrementally decoding.
+
+        Returns:
+            Tensor: output tensor with shape [b x s x v]
+
+        Notation used for tensor shapes:
+            - b: batch size
+            - s: sequence length
+            - v: vocab size
+            - d: embed dim
+        """
+        # input tensor of shape [b, s]
+        bsz, seq_len = tokens.shape
+
+        # shape: [b, s, d]
+        h = self.tok_embeddings(tokens)
+
+        # TODO: Fix the masking logic to not rely on checking kv_cache
+        if seq_len > 1 and self.layers[0].attn.kv_cache is not None:
+            mask = torch.full(
+                (1, 1, seq_len, seq_len), float("-inf"), device=tokens.device
+            )
+            mask = torch.triu(mask, diagonal=curr_pos + 1)
```

Comment:
Once you rebase, this logic will need to be updated. See how we setup the mask on main